DotnetFundamentals/Logger/Logger.cs

```diff
@@ -0,0 +1,139 @@
+             .SetBasePath(Directory.GetParent(workingDirectory).Parent.Parent.FullName)
+             .AddJsonFile("appsettings.json");
+
+            var config = builder.Build();
+
+            CreateDictionary(config);
+        }
+
+        public void Warning(string message)
+        {
+            InvokeMethod("Warning", message);
+        }
+
+        public void Info(string message)
+        {
+            InvokeMethod("Info", message);
+        }
+
+        public void Error(string message)
+        {
+            InvokeMethod("Error", message);
+        }
+
+        public void Error(Exception ex)
+        {
+            if (dependencyOnLevel.ContainsKey("Error"))
+            {
+                foreach (ILogger instance in dependencyOnLevel["Error"])
+                {
+                    instance.Error(ex);
+                }
+            }
+            else
+            {
+                var instance = new ConsoleLogger();
+                instance.Error(ex);
+            }
+        }
+
+        private void CreateDictionary(IConfigurationRoot config)
+        {
+            List<ILogger> Loggers = new List<ILogger>();
+            Dictionary<ILogger, string[]> LoggersMapping = new Dictionary<ILogger, string[]>();
+            var configuration = config.GetSection("Logging").Get<Dictionary<string, string[]>>();
+            var type = typeof(ILogger);
+            var types = AppDomain.CurrentDomain.GetAssemblies()
+                .SelectMany(s => s.GetTypes())
+                .Where(p => type.IsAssignableFrom(p));
+
+            foreach (string logger in configuration.Keys)
+            {
+                foreach (Type t in types)
+                {
+                    if (logger == t.Name)
+                    {
+                        Loggers.Add((ILogger)Activator.CreateInstance(t));
+                        LoggersMapping.Add(Loggers[Loggers.Count - 1], configuration[logger]);
+                    }
+                }
+            }
+
+            foreach (ILogger instance in LoggersMapping.Keys)
+            {
+                foreach (string instanceLevel in LoggersMapping[instance])
+                {
+                    if (dependencyOnLevel.ContainsKey(instanceLevel))
+                    {
+                        dependencyOnLevel[instanceLevel].Add(instance);
+                    }
+                    else
+                    {
+                        dependencyOnLevel.Add(instanceLevel, new List<ILogger> { instance });
+                    }
+                }
+            }
+        }
+
+        private void InvokeMethod(string level, string message)
+        {
+            if (dependencyOnLevel.ContainsKey(level))
+            {
+                foreach (ILogger instance in dependencyOnLevel[level])
+                {
+                    Type t = instance.GetType();
+                    MethodInfo method = t.GetMethod(level);
+                    method.Invoke(instance, new object[] { message });
```

Comment:
its not a good idea to use reflection in every logging method from performance perspective

Reply:
you can pass `Action<ILogger>` as an argument instead

Reply:
I didn't quite understand how to use Action<ILogger> in that situation. Could you please show me some examples?
I found another solution to this low-performance problem. You can check it.